ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/State/Infra.hs

```diff
@@ -170,32 +170,41 @@ reconstructSummary (History.Shape shape) transition (HardForkState st) =
     go :: Exactly xs' EraParams
        -> Telescope (K Past) (Current f) xs'
        -> NonEmpty xs' EraSummary
+    go ExactlyNil              t                   = case t of {}
     go (ExactlyCons params ss) (TS (K Past{..}) t) =
         NonEmptyCons (EraSummary pastStart (EraEnd pastEnd) params) $ go ss t
-    go (ExactlyCons params ExactlyNil) (TZ Current{..}) =
-        -- The current era is the last. We assume it lasts until all eternity.
-        NonEmptyOne (EraSummary currentStart EraUnbounded params)
-    go (ExactlyCons params (ExactlyCons nextParams _)) (TZ Current{..}) =
+    go (ExactlyCons params ss) (TZ Current{..}) =
         case transition of
           TransitionKnown epoch ->
             -- We haven't reached the next era yet, but the transition is
             -- already known. The safe zone applies from the start of the
             -- next era.
             let currentEnd = History.mkUpperBound params currentStart epoch
                 nextStart  = currentEnd
-            in NonEmptyCons EraSummary {
-                   eraStart  = currentStart
-                 , eraParams = params
-                 , eraEnd    = EraEnd currentEnd
-                 }
-             $ NonEmptyOne EraSummary {
-                   eraStart  = nextStart
-                 , eraParams = nextParams
-                 , eraEnd    = applySafeZone
-                                 nextParams
-                                 nextStart
-                                 (boundSlot nextStart)
-                 }
+            in case ss of
+              ExactlyCons nextParams _ ->
+                  NonEmptyCons EraSummary {
+                      eraStart  = currentStart
+                    , eraParams = params
+                    , eraEnd    = EraEnd currentEnd
+                    }
+                $ NonEmptyOne EraSummary {
+                      eraStart  = nextStart
+                    , eraParams = nextParams
+                    , eraEnd    = applySafeZone
+                                    nextParams
+                                    nextStart
+                                    (boundSlot nextStart)
+                    }
+              ExactlyNil               ->
+                -- HOWEVER, this code doesn't know what that era is! This can
+                -- arise when a node has not updated its code despite an
+                -- imminent hard fork.
+                NonEmptyOne EraSummary {
```

Comment:
This seems an interesting case - how do we know about the hard fork before the node is updated? We could know about the imminent PV bump, but a PV bump does not imply a hard fork. We only relate a PV bump to a hard fork through declarations in code.

Reply:
Ah, OK. I've looked more carefully now at how `TransitionKnown` _actually_ arises.

First step backwards from my focus on`reconstructSummary` ("the HFC part"):

https://github.com/input-output-hk/ouroboros-network/blob/dfa2697c2bcdefdc0a97db7f988bcf90aaff47b5/ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/State.hs#L112-L145

Second step ("the `ShelleyBlock` part"): 

https://github.com/input-output-hk/ouroboros-network/blob/dfa2697c2bcdefdc0a97db7f988bcf90aaff47b5/ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs#L175-L187

https://github.com/input-output-hk/ouroboros-network/blob/dfa2697c2bcdefdc0a97db7f988bcf90aaff47b5/ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs#L207-L217

Third step ("the `protocolInfoCardano` part"):

https://github.com/input-output-hk/ouroboros-network/blob/dfa2697c2bcdefdc0a97db7f988bcf90aaff47b5/ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs#L940-L961

https://github.com/input-output-hk/ouroboros-network/blob/dfa2697c2bcdefdc0a97db7f988bcf90aaff47b5/ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs#L632-L638

https://github.com/input-output-hk/ouroboros-network/blob/dfa2697c2bcdefdc0a97db7f988bcf90aaff47b5/ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs#L659-L665

https://github.com/input-output-hk/ouroboros-network/blob/dfa2697c2bcdefdc0a97db7f988bcf90aaff47b5/ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs#L683-L689

https://github.com/input-output-hk/ouroboros-network/blob/dfa2697c2bcdefdc0a97db7f988bcf90aaff47b5/ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs#L707-L713

https://github.com/input-output-hk/ouroboros-network/blob/dfa2697c2bcdefdc0a97db7f988bcf90aaff47b5/ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Node.hs#L536-L555

Fourth step ("the `cardano-node` part"):

https://github.com/input-output-hk/cardano-node/blob/0b32f893a170866deca50dc9ef303c0a11ba37f0/cardano-node/src/Cardano/Node/Protocol/Cardano.hs#L217-L284

I think what you're observing is that:

* We wouldn't reach the `TransitionKnown & ExactlyNil` case of `reconstructSummary.go` because code with the last known era `X` would not even have a `ProtocolTransitionParamsShelleyBased` argument for the `protocolInfoCardano` -- unless something had gone catastrophically off the rails in our PR review process.

I would agree with that, with a couple caveats:

* I agree because of the `ShelleyBlock` instance and the pattern in `protocolInfoCardano` -- but that's "by convention".

* Moreover, `reconstructSummary` is more abstract than `ShelleyBlock`.

I think, at the very least, I should write a comment in this case of `reconstructSummary.go` explaining (as in this comment) why we think it's impossible for `CardanoBlock`.

Reply:
I've force pushed a commit that adds a corresponding paragraph to the comment.

Reply:
That's a very comprehensive explanation! I agree with you that it makes sense to both leave this case in as-is and explain in comments why it's a rare case to see.